File: 프로그래머스/2/42626. 더 맵게/더 맵게.py
import heapq

def solution(scoville, k):
    heapq.heapify(scoville) # min_heap

    count = 0
    # heap peek가 k 이상일 때까지
    while scoville[0] < k:
        if len(scoville) == 1:
            return -1
        
        min_scov = heapq.heappop(scoville)
        next_min_scov = heapq.heappop(scoville)
        
        heapq.heappush(scoville, min_scov + next_min_scov * 2)
        
        count += 1
        
    return count

    
    

File: 프로그래머스/2/42626. 더 맵게/더 맵게.java
// min heap

// 만들 수 없는 경우 -1

import java.util.*;
import java.util.stream.*;

class Solution {
    public int solution(int[] scovilles, int K) {
        
        Queue<Integer> minHeap = new PriorityQueue<>();
        for (int i = 0; i < scovilles.length; i++) {
            minHeap.add(scovilles[i]);
        }
        
        int answer = 0;
        while (minHeap.size() >= 2 && minHeap.peek() < K) {
            int first = minHeap.remove();
            int second = minHeap.remove();
            
            int sum = first + second * 2;
            minHeap.add(sum);
            answer += 1;
        }
        
        if (minHeap.peek() < K) return -1;
        
        return answer;
    }
}